internal/provider/resource_user.go

```diff
@@ -5,143 +5,242 @@ import (
+			"name": schema.StringAttribute{
 				Description: "The name to be assigned to the user",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"display_name": {
-				Description: "The display name to be assigned to the user",
-				Type:        schema.TypeString,
+			"display_name": schema.StringAttribute{
+				Description: "The display name to be assigned to the user (optional)",
 				Optional:    true,
 			},
-			"password": {
+			"password": schema.StringAttribute{
 				Description: "The password to be assigned to the user",
-				Type:        schema.TypeString,
 				Required:    true,
+				Sensitive:   true,
+			},
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 			},
 		},
 	}
 }
 
-func resourceUserCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+// Configure enables provider-level data or clients to be set in the
+// provider-defined DataSource type. It is separately executed for each
+// ReadDataSource RPC.
+func (r *userResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
 
-	var diags diag.Diagnostics
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
 
-	name := d.Get("name").(string)
-	displayName := d.Get("display_name").(string)
-	password := d.Get("password").(string)
+	r.client = client
+}
+
+// Create is called when the provider must create a new resource. Config
+// and planned state values should be read from the
+// CreateRequest and new state values set on the CreateResponse.
+func (r *userResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data userResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	_, err := client.Users.CreateUser(juju.CreateUserInput{
-		Name:        name,
-		DisplayName: displayName,
-		Password:    password,
+	_, err := r.client.Users.CreateUser(juju.CreateUserInput{
+		Name:        data.Name.ValueString(),
+		DisplayName: data.DisplayName.ValueString(),
+		Password:    data.Password.ValueString(),
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create user, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("created user resource %q", data.Name))
```

Comment:
Interesting, wasn't aware of `tflog` 👍

Reply:
We need to change all of our logging over to this as we go.